plugins/webp-uploads/settings.php

```diff
@@ -204,6 +204,9 @@ function webp_uploads_use_picture_element_callback(): void {
 		#webp_uploads_picture_element_fieldset.disabled p {
 			opacity: 0.7;
 		}
+		#webp_uploads_picture_element_fieldset.disabled #webp_uploads_use_picture_element_label {
+			pointer-events: none;
```

Comment:
That works! However, it will still be toggled with the keyboard, right?

Anyway, it doesn't really matter if the disabled checkbox can or can't be changed.

Reply:
I don't understand this. Wouldn't toggling with the keyboard still be a problem as it would allow changing the value (since the checkbox is not technically disabled)?

I assume the reason we don't use the `disabled` attribute is that we still need to send the data with the form submission? If so, maybe we should conditionally add `readonly`, to ensure the value cannot be changed while it's "disabled"?

Reply:
> Wouldn't toggling with the keyboard still be a problem as it would allow changing the value (since the checkbox is not technically disabled)?

Yes. IMO, this is just a UX enhancement though. Also accounting for keyboard toggling the checkbox seems overkill. It's not like it's a problem if someone successfully checks or unchecks this "disabled" checkbox, since the state is ignored. The visual `disabled` treatment is just to let them know it doesn't have an effect, which the associated admin notice also makes clear.

This being said, to improve a11y then perhaps setting the `aria-disabled=true` attribute should be done to make this crystal clear.  I've done this in 25d2c02fdc2bfbf48d7751afece024c023d7b03f.

> I assume the reason we don't use the `disabled` attribute is that we still need to send the data with the form submission? 

That is correct. I also noticed a bug where even though even though the `webp_uploads_use_picture_element` checkbox is checked, when the `perflab_generate_webp_and_jpeg` is not checked and the form is submitted, the resulting updated page will still show that the `webp_uploads_use_picture_element` checkbox is unchecked. This is in spite of the fact that the underlying DB option is still `webp_uploads_use_picture_element`. The reason for this is that the `webp_uploads_is_picture_element_enabled()` helper was being used instead of `get_option( 'webp_uploads_use_picture_element' )`, and the helper will always return false if the `perflab_generate_webp_and_jpeg` option is not `true`. I've fixed this in 8c8a294f4a8b65c5ba3b48b29abac38fe50249ea (and 97dc62779a7022bede1ef61a60e20a685c05c532).

> If so, maybe we should conditionally add `readonly`, to ensure the value cannot be changed while it's "disabled"?

Unfortunately, `readonly` doesn't prevent a checkbox from being toggled: https://codepen.io/westonruter/pen/ExBbzPx

Reply:
> It's not like it's a problem if someone successfully checks or unchecks this "disabled" checkbox, since the state is ignored.

But I thought it isn't really ignored, since the checkbox isn't actually disabled, but only has the `disabled` class. So as far as I understand, it'll send the value to the backend, which then IMO makes it strange if you can change it (even though that probably won't have any real behavior implications).

> Unfortunately, `readonly` doesn't prevent a checkbox from being toggled: https://codepen.io/westonruter/pen/ExBbzPx

Ah, didn't know that, that's too bad. The other alternative to solve this would be to _actually_ disable, while temporarily injecting an `input[type="hidden"]` that's just there to continue sending the previous value back to the backend - since without sending it, WordPress would update the option to `null`.

Reply:
> But I thought it isn't really ignored, since the checkbox isn't actually disabled, but only has the `disabled` class.

Sorry, by ignored I mean that `webp_uploads_is_picture_element_enabled()` will always return false when the `perflab_generate_webp_and_jpeg` option is `false`, regardless of whether the `webp_uploads_use_picture_element` option is true or false. 

> So as far as I understand, it'll send the value to the backend, which then IMO makes it strange if you can change it (even though that probably won't have any real behavior implications).

So yes, technically you could change it, but it won't have any behavioral effect.

> The other alternative to solve this would be to _actually_ disable, while temporarily injecting an `input[type="hidden"]` that's just there to continue sending the previous value back to the backend - since without sending it, WordPress would update the option to `null`.

That's right. But this seems like maybe overkill.

Reply:
> > The other alternative to solve this would be to _actually_ disable, while temporarily injecting an `input[type="hidden"]` that's just there to continue sending the previous value back to the backend - since without sending it, WordPress would update the option to `null`.
> 
> That's right. But this seems like maybe overkill.

I think we should consider doing that. Both the current approach and this one are sort of hacky. But with what I'm proposing, I think we're on the safer side when it comes to UX and accessibility. Better to actually disable a checkbox that we want disabled than not disable it, but try to make it seem disabled.

The hidden input allows us to solve the disabled checkbox UX in a clean way.

Reply:
Actually, thinking about it further: Why don't we simply add `disabled` to the checkbox than doing all these workarounds? If the checkbox is disabled, nothing will be sent to the backend. This will set the `webp_uploads_use_picture_element` option to `null` - which is the same as an unchecked checkbox - which is actually what we want: If the checkbox is disabled, you _can't_ use `picture` element, so it makes sense for the option value to be `false`-y.

Reply:
Because let's say you had PICTURE element enabled, but then you temporarily turned off JPEG fallback. When you go back to turn on JPEG fallback, you'd then find that your PICTURE elements aren't automatically being served again. Keeping a non-disabled `input` for `webp_uploads_use_picture_element` ensures that the option isn't mutated unnecessarily. The `webp_uploads_is_picture_element_enabled()` helper function takes care of whether or not you can use PICTURE element based on the state of the two checkboxes.

Reply:
Okay, that makes sense. In that case though, I think my previous suggestion of using the `disabled` attribute on the checkbox and temporarily injecting `input[type="hidden"]` is a cleaner solution.

Reply:
Implemented in 42b1129b545dfce594e79516363be2ba284ffceb. I found it simpler to not have to inject/remove the hidden input, but to just leave it all the time.

[Screen recording 2024-08-14 10.44.47.webm](https://github.com/user-attachments/assets/10d1ac07-ec92-401f-8b3a-14b4f3c950bc)

Reply:
Thanks, @westonruter, for updating the PR. The latest changes look good to me.

Clever way to exclude the checkbox in form submission.